src/frontend/src/optimizer/plan_expr_visitor/strong.rs

```diff
@@ -0,0 +1,306 @@
+
+#[derive(Default)]
+pub struct Strong {
+    null_columns: FixedBitSet,
+}
+
+impl Strong {
+    fn new(null_columns: FixedBitSet) -> Self {
+        Self { null_columns }
+    }
+
+    // Returns whether the analyzed expression will definitely return null if
+    // all of a given set of input columns are null.
+    #[allow(dead_code)]
+    pub fn is_null(expr: &ExprImpl, null_columns: FixedBitSet) -> bool {
+        let strong = Strong::new(null_columns);
+        strong.is_null_visit(expr)
+    }
+
+    fn is_input_ref_null(&self, input_ref: &InputRef) -> bool {
+        self.null_columns.contains(input_ref.index())
+    }
+
+    fn is_null_visit(&self, expr: &ExprImpl) -> bool {
+        match expr {
+            ExprImpl::InputRef(input_ref) => self.is_input_ref_null(input_ref),
+            ExprImpl::Literal(literal) => literal.get_data().is_none(),
+            ExprImpl::FunctionCall(func_call) => self.is_null_function_call(func_call),
+            ExprImpl::FunctionCallWithLambda(_) => false,
+            ExprImpl::AggCall(_) => false,
+            ExprImpl::Subquery(_) => false,
+            ExprImpl::CorrelatedInputRef(_) => false,
+            ExprImpl::TableFunction(_) => false,
+            ExprImpl::WindowFunction(_) => false,
+            ExprImpl::UserDefinedFunction(_) => false,
+            ExprImpl::Parameter(_) => false,
+            ExprImpl::Now(_) => false,
+        }
+    }
+
+    fn is_null_function_call(&self, func_call: &FunctionCall) -> bool {
+        match func_call.func_type() {
+            // NOT NULL: This kind of expression is never null. No need to look at its arguments, if it has any.
+            ExprType::IsNull
+            | ExprType::IsNotNull
+            | ExprType::IsDistinctFrom
+            | ExprType::IsNotDistinctFrom
+            | ExprType::IsTrue
+            | ExprType::IsNotTrue
+            | ExprType::IsFalse
+            | ExprType::IsNotFalse => false,
+            // ANY: This kind of expression is null if and only if at least one of its arguments is null.
+            ExprType::Not
+            | ExprType::Equal
+            | ExprType::NotEqual
+            | ExprType::LessThan
+            | ExprType::LessThanOrEqual
+            | ExprType::GreaterThan
+            | ExprType::GreaterThanOrEqual
+            | ExprType::Like
+            | ExprType::Add
+            | ExprType::AddWithTimeZone
+            | ExprType::Subtract
+            | ExprType::Multiply
+            | ExprType::Modulus
+            | ExprType::Divide
+            | ExprType::Cast
+            | ExprType::Trim
+            | ExprType::Ltrim
+            | ExprType::Rtrim
+            | ExprType::Ceil
+            | ExprType::Floor
+            | ExprType::Extract
+            | ExprType::Greatest
+            | ExprType::Least => self.any_null(func_call),
+            ExprType::And | ExprType::Or | ExprType::Coalesce => self.all_null(func_call),
```

Comment:
What if `false or (nullable expression)` / `true and (nullable expression)`?
In this case `all_null` will return `false`, but actually the above expressions will evaluate to `null`, and the result of `is_null` should be `true`.
We might need add special corner case check for `And` and `Or`.

Reply:
I think this case should be handled by constant folding.

Reply:
I see, then this make sense to me.